src/rust/engine/process_execution/src/cache.rs

```diff
@@ -19,13 +24,20 @@ pub struct CommandRunner {
 }
 
 impl crate::CommandRunner for CommandRunner {
+  fn get_compatible_request(
+    &self,
+    req: &MultiPlatformExecuteProcessRequest,
+  ) -> Option<ExecuteProcessRequest> {
+    self.underlying.get_compatible_request(req)
+  }
+
   // TODO: Maybe record WorkUnits for local cache checks.
   fn run(
     &self,
-    req: ExecuteProcessRequest,
+    req: MultiPlatformExecuteProcessRequest,
     workunit_store: WorkUnitStore,
   ) -> BoxFuture<FallibleExecuteProcessResult, String> {
-    let digest = try_future!(self.digest(&req));
+    let digest = self.digest(req.clone());
```

Comment:
I'm not sold on these semantics... It feels like this should be "return cache hit if _any_ of the platform-specific versions were a cache hit", where the semantics here currently are "return cache hit if _all_ of the platform-specific versions have been seen together before"?

If I do a build which:

On linux, /bin/echo hello
Or on mac, /bin/echo hi

And then do a build:
On linux, /bin/echo hello

I want to hit the cache, right?

I think we want to instead do something like:
```rust
let digests: Vec<_> = self.digest(&req);
let futures = futures::future::join_all(digests.into_iter().map(|digest| self.lookup(digest)).collect::<Vec<_>>());
futures.and_then(|maybe_hits| {
  let hits: Option<FallibleExecuteProcessResult> = maybe_hits.filter_map(|maybe_hit| maybe_hit).next();
  ...
})
```

Open question as to whether we should write to the cache for _each_ of the digests, or just for the one that actually ran...

Reply:
I still think its okay to cache at the level of "have we seen all these requests together" because rules will always emit the same MultiPlatformEPR's (for a given options config), so seeing the same request combo in the cache key amounts to caching at the rule level, which is what we want right? It seems like an anti-pattern to have two rules that emit the same EPRs, but that is the only case where caching individual requests within a MultiPlatformEPR would result in additional cache hits. If that isn't correct and we expect multiple rules to yield the same ExecuteProcessRequests, then it could make sense to cache individual request.

>Open question as to whether we should write to the cache for each of the digests, or just for the one that actually ran...

If we do change to your suggestion of a cache lookup per sub-request in a MPEPR, then I think we should write the result to the cache for *each* request in the MPEPR. I am hesitant to add a mechanism that feeds information about which request actually ran back to the caller from a CommandRunner because it would require such a large change to the way the command runners are structured.

What the echo example gets wrong is that those requests produce different output depending on platform. The output shouldn't be used interchangeably in any case. I talked with @stuhood about this issue for a while and we agreed that rules should be responsible for making sure that *if* they output multiple requests for cross-platform support that the output is interchangeable, i.e all requests in a MultiPlatformEPR are configured to to produce the same output regardless of which one is selected based on the available platform. Assuming that is true then we can use whatever output was produced for a MPEPR as the cache value for any request in the MPEPR, if we choose to cache individually.

Reply:
> rules will always emit the same MultiPlatformEPR's (for a given options config)

But why should we throw away possible cache hits just because you ran pants twice with two different flags which don't change the output? If I'm not connected to the internet, so run offline just locally, then connect to the internet and can use either local or remoting, I'd still like to have the cache hit...

Reply:
CLI args would not affect where a MPEPR (pheew, mouthful) gets created: only code changes. Any code path that needs different per-platform behaviour needs to say so by using MPEPR.

Reply:
>Any code path that needs different per-platform behaviour needs to say so by using MPEPR

To correct myself / clarify. For now we don't have options to constrain target platforms so you're right, options wont affect the MPEPR output. In the future when target platform is constrainable that flag could change the MPEPR. Rules can't emit all possible (exec, target) platform combos always, because that would break the assumption that ll requests in a MPEPR produce the same output.

>But why should we throw away possible cache hits just because you ran pants twice with two different flags which don't change the output? If I'm not connected to the internet, so run offline just locally, then connect to the internet and can use either local or remoting, I'd still like to have the cache hit...

I think my comment was slightly misleading because I was referring to a future situation where we constrain the target platform. This example should work under the current setup.    The flow would look something like this:
* rule emits an MPEPR with multiple requests.
* (command runner is the local runner) command_runner::get_compatible_requests returns  the request which can run on the local platform.
* request is missing in the cache so the local command runner is used to actually run the request.
* result is cached under the composite key.
* (user connects to web and turns on remoting / speculation) previously run rule emits the same MPEPR
* MPEPR is still compatible, (perhaps now with the remote cluster), command_runner::run is called
* cache computes the same composite key, and previously computed result is returned.